File: ttnn/ttnn/operations/conv1d.py
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

from typing import Tuple, Union, Dict, Optional
import torch
import warnings
import math
import ttnn

Conv1dConfig = ttnn._ttnn.operations.conv.Conv2dConfig


@ttnn.register_python_operation(name="ttnn.conv1d")
def Conv1d(
    *,
    input_tensor: ttnn.Tensor,  # may or may not be sharded
    weight_tensor: ttnn.Tensor,
    device: ttnn.Device,
    in_channels: int,
    out_channels: int,
    batch_size: int,
    input_length: int,
    kernel_size: int,
    stride: int = 1,
    padding: int = 0,
    dilation: int = 1,
    groups: int = 1,
    bias_tensor: ttnn.Tensor = None,
    conv_config: Conv1dConfig = None,  # config overrides by user
    compute_config: ttnn.DeviceComputeKernelConfig = None,
    return_output_dim=False,
    return_weights_and_bias=False,
) -> Tuple[ttnn.Tensor, int, int, ttnn.Tensor, ttnn.Tensor]:
    """
    Applies a 1D convolution over an input signal composed of several input planes. Implemented as a 2D Convolution of input height 1 and input width as input_length.

        :param ttnn.Tensor input_tensor:  The input tensor. This must be in the format [N, H, W, C]. It can be on host or device.
        :param ttnn.Tensor weight_tensor: The weight tensor. The weights can be passed in the same format as PyTorch, [out_channels, in_channels, kernel_height, kernel_width]. The op w
        :param ttnn.Tensor, None bias_tensor:   Optional bias tensor. Default: None
        :param ttnn.IDevice device:  The device to use.
        :param int: in_channels:  Number of input channels.
        :param int: out_channels:  Number of output channels.
        :param int: batch_size:  Batch size.
        :param int: input_height:  Height of the input tensor.
        :param int: input_width:  Width of the input tensor.
        :param tuple[int  , int] kernel_size: Size of the convolving kernel.
        :param tuple[int, int] stride: Stride of the cross-correlation.
        :param tuple[int, int] or tuple[int, int, int, int]) padding: Zero-padding added to both sides of the input. [pad_height, pad_width] or [pad_top, pad_bottom, pad_left, pad_right].
        :param tuple[int, int] dilation: Spacing between kernel elements.
        :param int groups:  Number of blocked connections from input channels to output channels.
        :param ttnn.Conv2dConfig, None conv_config: Configuration for convolution. Default: None
        :param ttnn.DeviceComputeKernelConfig, None compute_config: Configuration for compute kernel. Default: None
        :param ttnn.MemoryConfig, None memory_config: Output Tensor's Memory Configuration. Default: None
        :param bool return_output_dim:  If true, the op also returns the height & width of the output tensor in [N, H, W, C] format,
        :param bool return_weights_and_bias:  If true, the op also returns the preprocessed weight and bias on device .

        :return: The output tensor, output height & width, and the preprocessed weights & bias.

        :rtype: [ttnn.Tensor]: The output tensor, when return_output_dim = False and return_weights_and_bias = False
        :rtype: [ttnn.Tensor, Tuple[int, int]]: The output tensor, and it's height & width, if return_output_dim = True
        :rtype: [ttnn.Tensor, Tuple[ttnn.Tensor, ttnn.Tensor]]: The output tensor, and it's height & width, if return_weights_and_bias = True
        :rtype: [ttnn.Tensor, Tuple[int, int], Tuple[ttnn.Tensor, ttnn.Tensor]]: The output tensor, and it's height & width, if return_output_dim = True and return_weights_and_bias = True

    """
    # Reshape the input and weight tensors to 4D for conv2d operation
    # Should be no-op as input_tensor is in RM layout
    if len(input_tensor.shape) != 4:
        input_tensor = ttnn.reshape(input_tensor, [batch_size, input_length, 1, in_channels])
    if len(weight_tensor.shape) != 4:
        weight_tensor = ttnn.reshape(weight_tensor, [out_channels, in_channels // groups, kernel_size, 1])

    (
        output_tensor_new,
        output_length_new,
        _,
        weight_tensor_on_dev_new,
        bias_tensor_on_dev_new,
    ) = ttnn._ttnn.operations.conv.conv2d(
        input_tensor=input_tensor,
        weight_tensor=weight_tensor,
        device=device,
        in_channels=in_channels,
        out_channels=out_channels,
        batch_size=batch_size,
        input_height=input_length,
        input_width=1,
        kernel_size=(kernel_size, 1),
        stride=(stride, 1),
        padding=(padding, 0),
        dilation=(dilation, 1),
        groups=groups,
        bias_tensor=bias_tensor,
        conv_config=conv_config,
        compute_config=compute_config,
    )

    if return_output_dim and return_weights_and_bias:
        return output_tensor_new, output_length_new, [weight_tensor_on_dev_new, bias_tensor_on_dev_new]
    elif return_weights_and_bias:
        return output_tensor_new, [weight_tensor_on_dev_new, bias_tensor_on_dev_new]
    elif return_output_dim:
        return output_tensor_new, output_length_new
    else:
        return output_tensor_new


__all__ = []

File: ttnn/cpp/ttnn/operations/conv/conv1d/conv1d.cpp
// SPDX-FileCopyrightText: © 2025 Tenstorrent USA, Inc.
//
// SPDX-License-Identifier: Apache-2.0

#include "ttnn/operations/conv/conv1d/conv1d.hpp"

#include <array>
#include <variant>

#include <tt_stl/assert.hpp>
#include "ttnn/operations/conv/conv_types.hpp"
#include "ttnn/operations/conv/conv2d/conv2d.hpp"
#include "ttnn/operations/core/core.hpp"

namespace ttnn {

Conv1dResult conv1d(
    const ttnn::Tensor& input_tensor,
    const ttnn::Tensor& weight_tensor,
    MeshDevice* device,
    uint32_t in_channels,
    uint32_t out_channels,
    uint32_t batch_size,
    uint32_t input_length,
    uint32_t kernel_size,
    uint32_t stride,
    std::variant<std::array<uint32_t, 2>, uint32_t> padding,
    uint32_t dilation,
    uint32_t groups,
    const std::optional<const DataType>& dtype,
    const std::optional<const ttnn::Tensor>& bias_tensor,
    const std::optional<const Conv1dConfig>& conv_config,
    const std::optional<const DeviceComputeKernelConfig>& compute_config,
    const std::optional<const MemoryConfig>& memory_config,
    const std::optional<const Conv1dSliceConfig>& slice_config,
    bool return_output_dim,
    bool return_weights_and_bias) {
    // reshape input tensor to 4D, if it is not already
    const ttnn::Tensor& input_tensor_4d =
        (input_tensor.logical_shape().rank() < 4)
            ? ttnn::reshape(input_tensor, Shape({batch_size, 1, input_length, in_channels}))
            : input_tensor;

    // Reinterpret the 3D conv1d weight [out_channels, in_channels/groups, kernel_size] as 4D
    // [.., 1, kernel_size], matching the [N, 1, input_length, C] input reshape and the {1, kernel_size}
    // kernel. conv2d's weight prep does this for the L1 path, but the DRAM slicing auto-shard path reads
    // weight.logical_shape()[3] (the kernel width) before weights are prepared, so it must be 4D up front.
    const ttnn::Tensor& weight_tensor_4d = (weight_tensor.logical_shape().rank() == 3)
                                               ? ttnn::reshape(
                                                     weight_tensor,
                                                     Shape(
                                                         {weight_tensor.logical_shape()[0],
                                                          weight_tensor.logical_shape()[1],
                                                          1,
                                                          weight_tensor.logical_shape()[2]}))
                                               : weight_tensor;

    // padding for conv2d based on conv1d padding
    std::variant<std::array<uint32_t, 2>, std::array<uint32_t, 4>> conv2d_padding;
    if (std::holds_alternative<uint32_t>(padding)) {
        conv2d_padding = std::array<uint32_t, 2>{0, std::get<uint32_t>(padding)};
    } else {
        std::array<uint32_t, 2> padding_lr = std::get<std::array<uint32_t, 2>>(padding);

        conv2d_padding = std::array<uint32_t, 4>{
            0,              // up
            0,              // down
            padding_lr[0],  // left
            padding_lr[1]   // right
        };
    };

    // Conv1d reshapes the input to [N, 1, input_length, C], so the height dimension is always 1.
    // DRAM slicing is therefore only meaningful along the width dimension (DRAM_WIDTH, i.e. input_length);
    // DRAM_HEIGHT would produce a single degenerate slice. Reject it only when requested explicitly.
    if (slice_config.has_value()) {
        TT_FATAL(
            slice_config->slice_type != ttnn::prim::Conv2dSliceConfig::SliceType::DRAM_HEIGHT,
            "Conv1D does not support DRAM_HEIGHT slicing because the convolution height is always 1. "
            "Use DRAM_WIDTH slicing (slices along input_length) or L1_FULL.");
    }
    // When no slice config is provided, forward nullopt so conv2d auto-routes by input location:
    // inputs already in L1 stay in L1, while DRAM/host inputs are width-sliced through DRAM (height is
    // always 1, so the auto-selected slice type is always DRAM_WIDTH). This lets long sequences that
    // would otherwise overflow L1 stream through DRAM by default instead of forcing L1_FULL. The DRAM
    // slicing path auto-determines a shard layout when conv_config.shard_layout is unset, so
    // shard_layout=None / auto_shard callers continue to work.
    const std::optional<const ttnn::prim::Conv2dSliceConfig>& effective_slice_config = slice_config;

    auto [output_tensor, output_dimensions, weights_and_bias] =
        std::get<static_cast<int>(ConvResultType::OUTPUT_DIM_WEIGHTS_AND_BIAS)>(ttnn::conv2d(
            input_tensor_4d,
            weight_tensor_4d,
            device,
            in_channels,
            out_channels,
            batch_size,
            1,             // input_height
            input_length,  // input_width
            std::array<uint32_t, 2>{1, kernel_size},
            std::array<uint32_t, 2>{1, stride},
            conv2d_padding,
            std::array<uint32_t, 2>{1, dilation},
            groups,
            dtype,
            bias_tensor,
            conv_config,
            compute_config,
            memory_config,
            effective_slice_config,
            true,
            true));

    if (return_output_dim && return_weights_and_bias) {
        return Conv1dResult(std::tuple(output_tensor, std::get<1>(output_dimensions), weights_and_bias));
    }
    if (return_output_dim) {
        return Conv1dResult(std::tuple(output_tensor, std::get<1>(output_dimensions)));
    }
    if (return_weights_and_bias) {
        return Conv1dResult(std::tuple(output_tensor, weights_and_bias));
    }
    return Conv1dResult(output_tensor);
}

}  // namespace ttnn
